src/vs/workbench/contrib/chat/browser/contrib/chatDynamicVariable.ts

```diff
@@ -0,0 +1,113 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { localize } from '../../../../../nls.js';
+import { URI } from '../../../../../base/common/uri.js';
+import { IDynamicVariable } from '../../common/chatVariables.js';
+import { basename } from '../../../../../base/common/resources.js';
+import { assertDefined } from '../../../../../base/common/types.js';
+import { IRange } from '../../../../../editor/common/core/range.js';
+import { Location } from '../../../../../editor/common/languages.js';
+import { PromptFileReference } from '../../common/promptFileReference.js';
+import { IFileService } from '../../../../../platform/files/common/files.js';
+import { IConfigurationService } from '../../../../../platform/configuration/common/configuration.js';
+
+/**
+ * Parse the `data` property of a reference as an `URI`.
+ * @throws if the `data` reference is `not defined` or an invalid `URI`.
+ */
+const parseUri = (data: IDynamicVariable['data']): URI | Location => {
+	assertDefined(
+		data,
+		`The reference must have a \`data\` property, got ${data}.`,
+	);
+
+	if (typeof data === 'string') {
+		return URI.parse(data);
```

Comment:
Is this a safe assumption? When is it a stringified URI?

Reply:
Not really sure, the type seem to be too overreaching and it's not clear why, so my assumption was that a `string` value here is a path. Can restrict it to `URI` and `Location` values only 🤔

Reply:
Actually not sure what is better in this case - throw immediately if value is a string, or try to parse and throw if it is not an URI. In the latter case we at least trying to do something about it 🤷

Reply:
The type has been stretched and needs refactoring. But all the instances of this come from our code so you can cover the cases that really exist and ignore the rest or cover them with asserts or whatever you want. But it's confusing to read this because other variables (not "dynamic") have a string value which would be your selected code or something like that.

Reply:
I'm confused by this and definitely don't know all the cases when the string value would be a valid URI, but assumed that the parse will throw if it is not. What would you suggest here besides trying to infer if the string is a valid URI manually? Should we reject the string values all together?

Reply:
And I don't quite understand `ChatDynamicVariable extends PromptFileReference` because not every dynamic variable has to be a "prompt file reference" right?

Reply:
Might be a good one to discuss in a call - there are 2 assumptions I've had here. (1) Is that a `chat dynamic variable` is synonymous with `file reference`, and hence (2) that `data` of it must always represent a valid `URI`, therefore we'd want to throw here if not since otherwise this is a logic error.